src/adapter/src/coord.rs

```diff
@@ -3522,6 +3540,55 @@ impl Coordinator {
             }
         }
     }
+
+    /// Prune all storage usage events from the [`MZ_STORAGE_USAGE_BY_SHARD`] table that are older
+    /// than `retention_period`.
+    ///
+    /// This method will read the entire contents of [`MZ_STORAGE_USAGE_BY_SHARD`] into memory
+    /// which can be expensive.
+    ///
+    /// DO NOT call this method outside of startup. The safety of reading at the current oracle read
+    /// timestamp and then writing at whatever the current write timestamp is (instead of
+    /// `read_ts + 1`) relies on the fact that there are no outstanding writes during startup.
+    ///
+    /// In practice, this method is always safe to call because group commit has builtin fencing,
+    /// and we never commit retractions to [`MZ_STORAGE_USAGE_BY_SHARD`] outside of this method. So
+    /// we don't have to worry about double retractions. Still, better safe than sorry.
```

Comment:
@aljoscha this is my wall of text explaining why this is safe, please read it carefully and let me know if you agree.

Reply:
Yes, my reasoning is:

Assumptions:
- group_commit confirms leadership, meaning it checks that we have not been fenced

1. by the time at which we call this method (prune), we have fenced out other processes from writing
2. we determine retractions
3. we try and apply them in group_commit

When 3. succeeds, no-one else can have tried sneaking in retractions.